rust/opendp-ffi/src/any.rs

```diff
@@ -0,0 +1,593 @@
+        // We have to clone sub_glue, because self is moved into the call.
+        self.sub_glue.clone()(self, rhs)
+    }
+}
+
+// TODO: If/when we remove the clone of the budget from make_adaptive_composition(), then remove Clone from AnyXXXDistance.
+#[derive(Clone, PartialEq)]
+pub struct AnyMetricDistance {
+    distance: AnyBoxClonePartialEq,
+    partial_cmp_glue: Glue<fn(&Self, &Self) -> Option<Ordering>>,
+}
+
+impl AnyMetricDistance {
+    pub fn new<Q: 'static + Clone + MetricDistance>(distance: Q) -> Self {
+        Self {
+            distance: AnyBoxClonePartialEq::new_clone_partial_eq(distance),
+            partial_cmp_glue: Glue::new(|self_: &Self, other: &Self| -> Option<Ordering> {
+                let self_ = self_.downcast_ref::<Q>().unwrap_assert("downcast of AnyMeasureDistance to constructed type will always work");
+                let other = other.downcast_ref::<Q>();
+                // FIXME: Do we want to have a FalliblePartialCmp for this?
+                other.map_or(None, |o| self_.partial_cmp(o))
+            }),
+        }
+    }
+}
+
+impl Downcast for AnyMetricDistance {
+    fn downcast<T: 'static>(self) -> Fallible<T> {
+        self.distance.downcast()
+    }
+    fn downcast_ref<T: 'static>(&self) -> Fallible<&T> {
+        self.distance.downcast_ref()
+    }
+}
+
+impl PartialOrd for AnyMetricDistance {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        (self.partial_cmp_glue)(self, other)
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub struct AnyMeasure {
+    measure: AnyBoxClonePartialEq,
+}
+
+impl AnyMeasure {
+    pub fn new<M: 'static + Measure>(measure: M) -> Self {
+        Self { measure: AnyBoxClonePartialEq::new_clone_partial_eq(measure) }
+    }
+}
+
+impl Downcast for AnyMeasure {
+    fn downcast<T: 'static>(self) -> Fallible<T> {
+        self.measure.downcast()
+    }
+    fn downcast_ref<T: 'static>(&self) -> Fallible<&T> {
+        self.measure.downcast_ref()
+    }
+}
+
+impl Default for AnyMeasure {
+    fn default() -> Self { unimplemented!("called AnyMeasure::default()") }
+}
+
+impl Measure for AnyMeasure {
+    type Distance = AnyMeasureDistance;
+}
+
+#[derive(Clone, PartialEq)]
+pub struct AnyMetric {
+    metric: AnyBoxClonePartialEq,
+}
+
+impl AnyMetric {
+    pub fn new<M: 'static + Metric>(metric: M) -> Self {
+        Self { metric: AnyBoxClonePartialEq::new_clone_partial_eq(metric) }
+    }
+}
+
+impl Downcast for AnyMetric {
+    fn downcast<T: 'static>(self) -> Fallible<T> {
+        self.metric.downcast()
+    }
+    fn downcast_ref<T: 'static>(&self) -> Fallible<&T> {
+        self.metric.downcast_ref()
+    }
+}
+
+impl Default for AnyMetric {
+    fn default() -> Self { unimplemented!("called AnyMetric::default()") }
+}
+
+impl Metric for AnyMetric {
+    type Distance = AnyMetricDistance;
+}
+
+type AnyFunction = Function<AnyDomain, AnyDomain>;
+
+pub trait IntoAnyFunctionExt {
+    fn into_any(self) -> AnyFunction;
+}
```

Comment:
This is in a separate trait, not an impl on Function, because they are in separate crates.